src/Workspaces/Core/Portable/Progress/CodeAnalysisProgress.cs

```diff
@@ -0,0 +1,149 @@
+    /// name="description"/> can also be provided to update the UI accordingly (see <see cref="Description"/>).
+    /// </summary>
+    /// <remarks>
+    /// Multiple items of work can be transitioned to be complete by passing an explicit value to <paramref
+    /// name="count"/>.
+    /// </remarks>
+    /// <param name="count">The number of items that were completed.  Defaults to <c>1</c> for the trivial case where
+    /// only a single item has been completed.  Must be greater than or equal to 1.</param>
+    /// <param name="description">Optional description to update the UI to.</param>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.CompleteItem());
+    /// </example>
+    public static CodeAnalysisProgress CompleteItem(int count = 1, string? description = null)
+        => new()
+        {
+            CompleteItemValue = count >= 1 ? count : throw new ArgumentOutOfRangeException(nameof(count)),
+            DescriptionValue = description,
+        };
+
+    /// <summary>
+    /// Indicates that all progress should be reset for the current operation. This is normally done when the code
+    /// action is performing some new phase and wishes for the UI progress bar to restart from the beginning.
+    /// </summary>
+    /// <remarks>
+    /// Currently internal as only roslyn needs this in the impl of our suggested action (we use a progress bar to
+    /// compute the work, then reset the progress to apply all the changes).  Could be exposed later to 3rd party code
+    /// if a demonstrable need is presented.
+    /// </remarks>
+    internal static CodeAnalysisProgress Clear()
+        => new() { ClearValue = true };
+}
+
+internal sealed class CodeAnalysisProgressTracker(Action<string?, int, int>? updateAction) : IProgress<CodeAnalysisProgress>
+{
+    private string? _description;
+    private int _completedItems;
+    private int _totalItems;
+
+    public CodeAnalysisProgressTracker()
+        : this(null)
+    {
+    }
+
+    public string? Description
+    {
+        get => _description;
+        set
+        {
+            _description = value;
+            Update();
+        }
+    }
+
+    public int CompletedItems => _completedItems;
+
+    public int TotalItems => _totalItems;
+
+    public void AddItems(int count)
+    {
+        Interlocked.Add(ref _totalItems, count);
+        Update();
+    }
+
+    public void CompleteItems(int count)
+    {
+        Interlocked.Add(ref _completedItems, count);
+        Update();
+    }
+
+    public void Clear()
+    {
+        _totalItems = 0;
+        _completedItems = 0;
+        _description = null;
+        Update();
+    }
+
+    private void Update()
+        => updateAction?.Invoke(_description, _completedItems, _totalItems);
+
+    public void Report(CodeAnalysisProgress value)
+    {
+        if (value.DescriptionValue != null)
+            this.Description = value.DescriptionValue;
```

Comment:
Learning opportunity for me, when is it preferable to use the "this" qualifier in Roslyn?

Reply:
i find naked `PascalCasesThingy = ...` very hard to read.  It looks like a type name.  So, when naked, i prefer a prefixed 'this' to indicate its' a member.

But we have no rule on it and you can do whatever you want.

Reply:
OK, I guess I was hoping for something a bit more concrete. It just looks odd to me when calls to Add/Complete Items uses the this qualifer here but the Update method is called above without the qualifier.